visualization/src/point_picking_event.cpp

```diff
@@ -167,49 +172,66 @@ pcl::visualization::PointPickingCallback::performSinglePick (
 /////////////////////////////////////////////////////////////////////////////////////////////
 int
 pcl::visualization::PointPickingCallback::performAreaPick (vtkRenderWindowInteractor *iren,
-                                                           std::vector<int> &indices) const
+                                                           CloudActorMapPtr cam_ptr,
+                                                           std::map<std::string, Indices>& cloud_indices) const
 {
   vtkAreaPicker *picker = static_cast<vtkAreaPicker*> (iren->GetPicker ());
   vtkRenderer *ren = iren->FindPokedRenderer (iren->GetEventPosition ()[0], iren->GetEventPosition ()[1]);
   picker->AreaPick (x_, y_, iren->GetEventPosition ()[0], iren->GetEventPosition ()[1], ren);
-  if (picker->GetDataSet ())
-  {
-    vtkPolyData* points = reinterpret_cast<vtkPolyData*> (picker->GetDataSet ());
-
-    // This is a naive solution till we fugure out where to add the GlobalIds at an earlier stage
-    if (!points->GetPointData ()->GetGlobalIds () ||
-        points->GetPointData ()->GetGlobalIds ()->GetNumberOfTuples () == 0)
-    {
-      vtkSmartPointer<vtkIdTypeArray> global_ids = vtkIdTypeArray::New ();
-      global_ids->SetNumberOfValues (picker->GetDataSet ()->GetNumberOfPoints ());
-      for (vtkIdType i = 0; i < picker->GetDataSet ()->GetNumberOfPoints (); ++i)
-        global_ids->SetValue (i,i);
 
-      points->GetPointData ()->SetGlobalIds (global_ids);
-    }
+  vtkProp3DCollection* props = picker->GetProp3Ds ();
+  if (!props)
+    return -1;
 
-    vtkPlanes* frustum = picker->GetFrustum ();
+  int pt_numb = 0;
+  vtkCollectionSimpleIterator pit;
+  vtkProp3D* prop;
+  for (props->InitTraversal (pit); (prop = props->GetNextProp3D (pit));)
+  {
+    vtkSmartPointer<vtkActor> actor = vtkActor::SafeDownCast (prop);
+    if (!actor)
+      continue;
 
-    vtkSmartPointer<vtkExtractGeometry> extract_geometry = vtkSmartPointer<vtkExtractGeometry>::New ();
-    extract_geometry->SetImplicitFunction (frustum);
+    vtkPolyData* pd = vtkPolyData::SafeDownCast (actor->GetMapper ()->GetInput ());
+    if(pd->GetPointData ()->HasArray ("Indices"))
+      pd->GetPointData ()->RemoveArray ("Indices");
 
-    extract_geometry->SetInputData (picker->GetDataSet ());
+    vtkSmartPointer<vtkIdTypeArray> ids = vtkSmartPointer<vtkIdTypeArray>::New ();
+    ids->SetNumberOfComponents (1);
+    ids->SetName ("Indices");
+    for(vtkIdType i = 0; i < pd->GetNumberOfPoints (); i++)
+      ids->InsertNextValue (i);
+    pd->GetPointData ()->AddArray (ids);
 
+    vtkSmartPointer<vtkExtractGeometry> extract_geometry = vtkSmartPointer<vtkExtractGeometry>::New ();
+    extract_geometry->SetImplicitFunction (picker->GetFrustum ());
+#if VTK_MAJOR_VERSION < 6
```

Comment:
Our minimum supported VTK version is 6.2

Reply:
Ok I remove îf clause